src/compiler/program.ts

```diff
@@ -158,24 +222,57 @@ namespace ts {
         let start = new Date().getTime();
 
         host = host || createCompilerHost(options);
+        
+        // initialize resolveModuleNameWorker only if noResolve is false
+        let resolveModuleNamesWorker: (moduleNames: string[], containingFile: string) => string[];
+        if (!options.noResolve) {
+            resolveModuleNamesWorker = host.resolveModuleNames;
+            if (!resolveModuleNamesWorker) {
+                Debug.assert(host.getModuleResolutionHost !== undefined);

```

Comment:
As we discussed, i think this should be directly on the host. this way you will get a compile-time error.